nvflare/app_common/job_launcher/process_launcher.py

```
@@ -25,41 +27,104 @@
 from nvflare.utils.job_launcher_utils import add_custom_dir_to_path, extract_job_image
 
 JOB_RETURN_CODE_MAPPING = {0: JobReturnCode.SUCCESS, 1: JobReturnCode.EXECUTION_ERROR, 9: JobReturnCode.ABORTED}
+_POSIX_SPAWN_SUPPORTED = hasattr(os, "posix_spawn") and os.name == "posix"
 
 
 class ProcessHandle(JobHandleSpec):
-    def __init__(self, process):
+    def __init__(self, process: Optional[subprocess.Popen] = None, pid: Optional[int] = None):
         super().__init__()
 
         self.process = process
+        self.pid = pid if pid is not None else (process.pid if process else None)
         self.logger = logging.getLogger(self.__class__.__name__)
+        self._return_code: Optional[int] = None
+
+        if self.pid is None:
+            raise ValueError("ProcessHandle requires either a process object or a pid.")
 
     def terminate(self):
-        if self.process:
-            try:
-                os.killpg(os.getpgid(self.process.pid), 9)
-                self.logger.debug("kill signal sent")
-            except:
-                pass
+        self._kill_process_group()
 
+        if self.process:
             self.process.terminate()
 
     def poll(self):
         if self.process:
-            return JOB_RETURN_CODE_MAPPING.get(self.process.poll(), JobReturnCode.EXECUTION_ERROR)
-        else:
-            return JobReturnCode.UNKNOWN
+            code = self.process.poll()
+            if code is None:
+                return JobReturnCode.UNKNOWN
+            return JOB_RETURN_CODE_MAPPING.get(code, JobReturnCode.EXECUTION_ERROR)
+
+        return self._poll_pid()
 
     def wait(self):
         if self.process:
             self.process.wait()
+            return
+
+        if self.pid is None:
+            return
+
+        if self._return_code is None:
+            try:
+                _, status = os.waitpid(self.pid, 0)
+                self._return_code = self._decode_status(status)
+            except ChildProcessError:
+                pass
```

are we swallow errors ?